crypto/sm2/sm2_crypt.c

```diff
@@ -0,0 +1,262 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 Ribose Inc. All Rights Reserved.
+ * Ported from Ribose contributions from Botan.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/sm2.h>
+#include <openssl/evp.h>
+#include <openssl/asn1.h>
+#include <openssl/asn1t.h>
+#include <string.h>
+
+typedef struct SM2_Ciphertext_st SM2_Ciphertext;
+DECLARE_ASN1_FUNCTIONS(SM2_Ciphertext)
+
+struct SM2_Ciphertext_st {
+    BIGNUM *C1x;
+    BIGNUM *C1y;
+    ASN1_OCTET_STRING *C3;
+    ASN1_OCTET_STRING *C2;
+};
+
+ASN1_SEQUENCE(SM2_Ciphertext) =
+{
+ASN1_SIMPLE(SM2_Ciphertext, C1x, BIGNUM),
+        ASN1_SIMPLE(SM2_Ciphertext, C1y, BIGNUM),
+        ASN1_SIMPLE(SM2_Ciphertext, C3, ASN1_OCTET_STRING),
+        ASN1_SIMPLE(SM2_Ciphertext, C2,
+                        ASN1_OCTET_STRING),} ASN1_SEQUENCE_END(SM2_Ciphertext)
+IMPLEMENT_ASN1_FUNCTIONS(SM2_Ciphertext)
+
+static int EC_field_size(const EC_GROUP *group)
+{
+    /* Is there some simpler way to do this? */
+    BIGNUM *p = BN_new();
+    BIGNUM *a = BN_new();
+    BIGNUM *b = BN_new();
```

Comment:
Should check all of these allocations for NULL, or use a BN_CTX and check the last allocation.

Reply:
Fixed